devices/usbc/ctrl.c

```diff
@@ -0,0 +1,400 @@
+	ctrl_common.dc->endptqh[1].dtd_next = 1;
+
+	ctrl_common.dc->endptqh[0].base = qh_addr + (32 * sizeof(dqh_t));
+	ctrl_common.dc->endptqh[0].size = 0x40;
+	ctrl_common.dc->endptqh[0].head = (dtd_t *)(ctrl_common.dc->endptqh + 32);
+	ctrl_common.dc->endptqh[0].tail = (dtd_t *)(ctrl_common.dc->endptqh + 32);
+
+	ctrl_common.dc->endptqh[1].base = qh_addr + (48 * sizeof(dqh_t));
+	ctrl_common.dc->endptqh[1].size = 0x10;
+	ctrl_common.dc->endptqh[1].head = (dtd_t *)(ctrl_common.dc->endptqh + 48);
+	ctrl_common.dc->endptqh[1].tail = (dtd_t *)(ctrl_common.dc->endptqh + 48);
+
+	*(ctrl_common.dc->base + endpointlistaddr) = qh_addr;
+	*(ctrl_common.dc->base + endptprime) |= 1;
+	*(ctrl_common.dc->base + endptprime) |= 1 << 16;
+
+
+	return ERR_NONE;
+}
+
+
+static dtd_t *ctrl_getDtd(int endpt, int dir)
+{
+	int qh = endpt * 2 + dir;
+	u32 base_addr;
+	dtd_t *ret;
+
+	base_addr = ((u32)ctrl_common.dc->endptqh[qh].head & ~((ctrl_common.dc->endptqh[qh].size * sizeof(dtd_t)) - 1));
+
+	ret = ctrl_common.dc->endptqh[qh].tail++;
+	ctrl_common.dc->endptqh[qh].tail = (dtd_t *)(base_addr | ((u32)ctrl_common.dc->endptqh[qh].tail & (((ctrl_common.dc->endptqh[qh].size) * sizeof(dtd_t)) - 1)));
+
+	return ret;
+}
+
+
+static int ctrl_buildDtd(dtd_t *dtd, u32 paddr, u32 size)
+{
+	int i = 0;
+	int tempSize = size;
+
+	if (size > USB_BUFFER_SIZE)
+		return -1;
+
+	dtd->dtd_next = 1;
+	dtd->dtd_token = size << 16;
+	dtd->dtd_token |= 1 << 7;
+
+	/* TODO: allow to use additional dtd within the same dQH */
+	while (tempSize > 0) {
+		dtd->buff_ptr[i++] = paddr;
+		tempSize -= 0x1000;
+		paddr += 0x1000;
+	}
+
+	return ERR_NONE;
+}
+
+
+dtd_t *ctrl_execTransfer(int endpt, u32 paddr, u32 sz, int dir)
+{
+	int shift;
+	u32 offs;
+	volatile dtd_t *dtd;
+
+	int qh = (endpt << 1) + dir;
+
+	dtd = ctrl_getDtd(endpt, dir);
+
+	ctrl_buildDtd((dtd_t *)dtd, paddr, sz);
+
+	shift = endpt + ((qh & 1) ? 16 : 0);
+	offs = (u32)dtd & (((ctrl_common.dc->endptqh[qh].size) * sizeof(dtd_t)) - 1);
+
+	ctrl_common.dc->endptqh[qh].dtd_next = (ctrl_common.dc->endptqh[qh].base + offs) & ~1;
+	ctrl_common.dc->endptqh[qh].dtd_token &= ~(1 << 6);
+	ctrl_common.dc->endptqh[qh].dtd_token &= ~(1 << 7);
+
+	while ((*(ctrl_common.dc->base + endptprime) & (1 << shift)))
+		;
+
+	*(ctrl_common.dc->base + endptprime) |= 1 << shift;
+
+	/* prime the endpoint and wait for it to prime */
+	while ((*(ctrl_common.dc->base + endptprime) & (1 << shift)))
+		;
+	*(ctrl_common.dc->base + endptprime) |= 1 << shift;
+	while (!(*(ctrl_common.dc->base + endptprime) & (1 << shift)) && (*(ctrl_common.dc->base + endptstat) & (1 << shift)))
+		;
+
+	/* wait to finish transaction */
+	while (DTD_ACTIVE(dtd) && !DTD_ERROR(dtd))
+		;
+
+	return (dtd_t *)dtd;
+}
+
+
+int ctrl_hfIrq(void)
+{
+	int endpt = 0;
+
+	if ((ctrl_common.dc->setupstat = *(ctrl_common.dc->base + endptsetupstat)) & 0x1) {
+		/* trip winre set */
+		while (!((ctrl_common.dc->setupstat >> endpt) & 1))
+			endpt++;
```

Comment:
Is there a chance that we end up with endpt = 32? This would cause bad memory access below.

Reply:
Usually host is responsible for initiating transmission with the device. So that, there is a check in line 300 whether setup transaction is received on control endpoint and then other endpoints are checked in the `while` loop.
I think that there is very little chance of that interrupts is received and no setup packet is transmitted on any of the endpoints. 

However, if we end up with`endpt > 15`, there would be definitely a bad memory access. For this reason it would be safer to check  `endpt` value before using it.

To sum up, thanks for comment ;)